File: online judge.kaner murillo/c++/10069.cpp
//https://onlinejudge.org/external/100/10069.pdf
//10069

#include<cstdio>
#include<iostream>
#include<string>

using namespace std;

int mem[10001][101];

int dp(int l, int r, string &cad1, string &cad2){

    int ans;
    if(mem[l][r]!=-1){
        ans = mem[l][r];
    }
    else{
        if(l == cad1.size() || r==cad2.size()){
            ans = 0;
            if(r==cad2.size()){
                ans = 1;
            }
        }
        else{
            ans = dp(l+1, r, cad1, cad2);
            if(cad1[l]==cad2[r]){
                ans += dp(l+1, r+1, cad1, cad2);
            }
        }
        mem[l][r] = ans;
    }

    return ans;
}

int main(){

    string cad1, cad2;
    int i, k, j, cases;

    cin>>cases;

    for(i = 0; i<cases; i++){
        cin>>cad1>>cad2;
        for(j = 0; j<=cad1.size(); j++){
            for(k = 0; k<=cad2.size(); k++){
                mem[j][k] = -1;
            }
        }
        printf("%i\n", dp(0, 0, cad1, cad2));
    }

    return 0;
}

File: online judge.kaner murillo/python/10069.py
#https://onlinejudge.org/external/100/10069.pdf
#10069

from sys import stdin
import sys

sys.setrecursionlimit(1000000)

def dp(i, j, cad1, cad2, mem):

    ans = None
    key = (i, j)
    if key in mem:
        ans = mem[key]

    else:
        
        if i == len(cad1) or j==len(cad2):
            ans = 0
            if j == len(cad2):
                ans = 1
        
        else:

            ans = dp(i+1, j, cad1, cad2, mem)

            if cad1[i] == cad2[j]:
                ans+=dp(i+1, j+1, cad1, cad2, mem)
            
        mem[key] = ans
    
    return ans

def main():

    cases = int(stdin.readline())

    for i in range(cases):
        
        cad1 = stdin.readline().strip()
        cad2 = stdin.readline().strip()
        mem = {}

        print(dp(0, 0, cad1, cad2, mem))

main()